### PR TITLE
Fix the bad assertion in ClasspathUtil

### LINKED ISSUE #23625: Potentially bad assertion in ClasspathUtil
During IntelliJ import tests, the ClasspathUtil is failing with an AssertionError:

<details>
<summary>Stacktrace</summary>

```text
Caused by: java.lang.AssertionError
	at org.gradle.internal.classloader.ClasspathUtil.getClasspathForResource(ClasspathUtil.java:136)
	at org.gradle.internal.classloader.TransformingClassLoader.findClass(TransformingClassLoader.java:64)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:587)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
	at kotlin.reflect.jvm.internal.impl.descriptors.runtime.components.ReflectKotlinClassFinder.<init>(ReflectKotlinClassFinder.kt:30)
	at kotlin.reflect.jvm.internal.impl.descriptors.runtime.components.RuntimeModuleData$Companion.create(RuntimeModuleData.kt:65)
	at kotlin.reflect.jvm.internal.ModuleByClassLoaderKt.getOrCreateModule(moduleByClassLoader.kt:58)
	at kotlin.reflect.jvm.internal.KDeclarationContainerImpl$Data$moduleData$2.invoke(KDeclarationContainerImpl.kt:36)
	at kotlin.reflect.jvm.internal.KDeclarationContainerImpl$Data$moduleData$2.invoke(KDeclarationContainerImpl.kt:33)
	at kotlin.reflect.jvm.internal.ReflectProperties$LazySoftVal.invoke(ReflectProperties.java:92)
	at kotlin.reflect.jvm.internal.ReflectProperties$Val.getValue(ReflectProperties.java:31)
	at kotlin.reflect.jvm.internal.KDeclarationContainerImpl$Data.getModuleData(KDeclarationContainerImpl.kt)
	at kotlin.reflect.jvm.internal.KClassImpl$Data$descriptor$2.invoke(KClassImpl.kt:50)
	at kotlin.reflect.jvm.internal.KClassImpl$Data$descriptor$2.invoke(KClassImpl.kt:47)
	at kotlin.reflect.jvm.internal.ReflectProperties$LazySoftVal.invoke(ReflectProperties.java:92)
	at kotlin.reflect.jvm.internal.ReflectProperties$Val.getValue(ReflectProperties.java:31)
	at kotlin.reflect.jvm.internal.KClassImpl$Data.getDescriptor(KClassImpl.kt)
	at kotlin.reflect.jvm.internal.KClassImpl.getDescriptor(KClassImpl.kt:182)
	at kotlin.reflect.jvm.internal.KClassImpl.getDescriptor(KClassImpl.kt:44)
	at kotlin.reflect.full.KClassifiers.createType(KClassifiers.kt:47)
	at kotlin.reflect.jvm.internal.ReflectionFactoryImpl.typeOf(ReflectionFactoryImpl.java:123)
	at kotlin.jvm.internal.Reflection.typeOf(Reflection.java:127)
	at org.jetbrains.kotlin.idea.gradleTooling.IdeaKotlinExtras.<clinit>(IdeaKotlinExtras.kt:52)
	at org.jetbrains.kotlin.idea.gradleTooling.IdeaKotlinExtras$Surrogate.readResolve(IdeaKotlinExtras.kt:42)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at java.base/java.io.ObjectStreamClass.invokeReadResolve(ObjectStreamClass.java:1190)
	at java.base/java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2266)
	at java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1733)
	at java.base/java.io.ObjectInputStream$FieldValues.<init>(ObjectInputStream.java:2606)
	at java.base/java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2457)
	at java.base/java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2257)
	at java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1733)
	at java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:509)
	at java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:467)
	at java.base/java.util.HashMap.readObject(HashMap.java:1552)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at java.base/java.io.ObjectStreamClass.invokeReadObject(ObjectStreamClass.java:1100)
	at java.base/java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2423)
	at java.base/java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2257)
	at java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1733)
	at java.base/java.io.ObjectInputStream$FieldValues.<init>(ObjectInputStream.java:2606)
	at java.base/java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2457)
	at java.base/java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2257)
	at java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1733)
	at java.base/java.io.ObjectInputStream$FieldValues.<init>(ObjectInputStream.java:2606)
	at java.base/java.io.ObjectInputStream.defaultReadObject(ObjectInputStream.java:647)
	at org.gradle.tooling.internal.adapter.ProtocolToModelAdapter$InvocationHandlerImpl.readObject(ProtocolToModelAdapter.java:381)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at java.base/java.io.ObjectStreamClass.invokeReadObject(ObjectStreamClass.java:1100)
	at java.base/java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2423)
	at java.base/java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2257)
	at java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1733)
	at java.base/java.io.ObjectInputStream$FieldValues.<init>(ObjectInputStream.java:2606)
	at java.base/java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2457)
	at java.base/java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2257)
	at java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1733)
	at java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:509)
	at java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:467)
	at java.base/java.util.HashMap.readObject(HashMap.java:1552)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at java.base/java.io.ObjectStreamClass.invokeReadObject(ObjectStreamClass.java:1100)
	at java.base/java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2423)
	at java.base/java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2257)
	at java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1733)
	at java.base/java.io.ObjectInputStream$FieldValues.<init>(ObjectInputStream.java:2606)
	at java.base/java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2457)
	at java.base/java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2257)
	at java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1733)
	at java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:509)
	at java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:467)
	at org.gradle.tooling.internal.provider.serialization.PayloadSerializer.deserialize(PayloadSerializer.java:76)
```
</details>

Attached debugger at failing position:
image

The conditoin
```java
schemeSpecificPart.substring(pos + 1).endsWith("/" + name)
```

would be true, however since the string 'name' is prefixed by '/META-INF/versions/9' the assertion fails.

### Expected Behavior
As disabling the assertion leads to the no user visible issues, I would expect no assertion to be thrown.

### Context
This affects Gradle Sync tests inside IntelliJ

### PR DESCRIPTION
**Pull request by Chaoba**
Signed-off-by: Yanshun Li <qqliys@126.com>

Fixes #23625

### Context

In the method `ClasspathUtil.getClasspathForResource`,  `equals` is used to assert the resource name, but it will fail if the name is prefixed.
So this pr uses `endsWith` to replace `equals`.

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation

### CODE REVIEW

File: subprojects/base-services/src/main/java/org/gradle/internal/classloader/ClasspathUtil.java [hunk 1]
```diff
@@ -133,7 +133,7 @@ public static File getClasspathForResource(URL resource, String name) {
                 String schemeSpecificPart = location.getRawSchemeSpecificPart();
                 int pos = schemeSpecificPart.indexOf("!");
                 if (pos > 0) {
-                    assert schemeSpecificPart.substring(pos + 1).equals("/" + name);
+                    assert schemeSpecificPart.substring(pos + 1).endsWith(name);
```
Comment (ljacomet) [inline subprojects/base-services/src/main/java/org/gradle/internal/classloader/ClasspathUtil.java:136]: I would keep the assertion consistent with the one on line 130 and check `.endsWith("/" + name)`

**Review by ljacomet — CHANGES_REQUESTED** [review]
Thank you for the PR, see minor improvement request below:

File: subprojects/base-services/src/main/java/org/gradle/internal/classloader/ClasspathUtil.java (hunk 1 above)
Comment (Chaoba) [reply subprojects/base-services/src/main/java/org/gradle/internal/classloader/ClasspathUtil.java]: Fixe.

**Comment by Chaoba** [comment]
@ljacomet Thank you for your comment. Please review it again.

**Review by ljacomet — APPROVED** [review]

**Comment by ljacomet** [comment]
@Chaoba Thanks for following up, this looks good to be merged

**Comment by ljacomet** [comment]
@bot-gradle test and merge

File: subprojects/base-services/src/test/groovy/org/gradle/internal/classloader/ClasspathUtilTest.groovy [hunk 2]
```diff
@@ -36,4 +37,27 @@ class ClasspathUtilTest extends Specification {
...
+
+    def "getClasspathForResource for jar scheme with resource name"() {
+        given:
+        def jarUrlWithResourceName = new URL('jar:file:/home/duke/duke.jar!/Test.class')
+
+        when:
+        def file = ClasspathUtil.getClasspathForResource(jarUrlWithResourceName, "Test.class")
+
+        then:
+        "/home/duke/duke.jar" == file.getPath()
```
Comment (ljacomet) [inline subprojects/base-services/src/test/groovy/org/gradle/internal/classloader/ClasspathUtilTest.groovy:49]: I missed that this assertion and the next one need to be platform agnostic.
Saw the following failure on Windows:
```
Condition not satisfied:
  
  "/home/duke/duke.jar" == file.getPath()
                        |  |    |
                        |  |    \home\duke\duke.jar
                        |  \home\duke\duke.jar
                        false
                        3 differences (84% similarity)
                        (/~)home(/~)duke(/~)duke.jar
                        (\\)home(\\)duke(\\)duke.jar
```

Applies to the one below as well.

**Review by ljacomet — CHANGES_REQUESTED** [review]
Was to quick on approving, CI found the following:

File: subprojects/base-services/src/test/groovy/org/gradle/internal/classloader/ClasspathUtilTest.groovy (hunk 2 above)
Comment (Chaoba) [reply subprojects/base-services/src/test/groovy/org/gradle/internal/classloader/ClasspathUtilTest.groovy]: Updated the test cases.  I don't have a Windows, only verified on Mac, sorry for that.

**Comment by ljacomet** [comment]
@bot-gradle test this

**Review by ljacomet — APPROVED** [review]
Thanks for the quick follow up, looks good now

**Comment by ljacomet** [comment]
@bot-gradle test and merge

**Pull request merged**